packages/cubejs-server-core/src/core/CompilerApi.js

```diff
@@ -481,39 +481,59 @@ export class CompilerApi {
       });
     };
 
-    return cubes
-      .map((cube) => ({
-        config: {
-          ...cube.config,
-          measures: cube.config.measures?.map(visibilityPatcherForCube(cube)),
-          dimensions: cube.config.dimensions?.map(visibilityPatcherForCube(cube)),
-          segments: cube.config.segments?.map(visibilityPatcherForCube(cube)),
-          hierarchies: cube.config.hierarchies?.map(visibilityPatcherForCube(cube)),
-        },
-      }));
+    const visibiliyMask = JSON.stringify(isMemberVisibleInContext, Object.keys(isMemberVisibleInContext).sort());
+    const visibilityMaskHash = crypto.createHash('md5').update(visibiliyMask).digest('hex').slice(0, 8);
+
+    return {
+      cubes: cubes
+        .map((cube) => ({
+          config: {
+            ...cube.config,
+            measures: cube.config.measures?.map(visibilityPatcherForCube(cube)),
+            dimensions: cube.config.dimensions?.map(visibilityPatcherForCube(cube)),
+            segments: cube.config.segments?.map(visibilityPatcherForCube(cube)),
+            hierarchies: cube.config.hierarchies?.map(visibilityPatcherForCube(cube)),
+          },
+        })),
+      visibilityMaskHash
+    };
+  }
+
+  mixInVisibilityMaskHash(compilerId, visibilityMaskHash) {
+    const uuidBytes = uuidParse(compilerId);
+    const hashBytes = Buffer.from(visibilityMaskHash, 'hex');
+    // Mix hash into the first 4 bytes and the last 4 bytes in a way that doesn't change the UUID version
+    for (let i = 0; i < 4; i++) {
+      // eslint-disable-next-line no-bitwise
+      uuidBytes[i] ^= hashBytes[i]; // First 4 bytes
```

Comment:
Let's not use XOR mixing, and not do this patching manually, please.

1. One can provide 16 random bytes to `uuid.v4` with `random` option. See [link](https://github.com/uuidjs/uuid?tab=readme-ov-file#uuidv4options-buffer-offset)
2. If we want to derive new pseudorandom value deterministically, simplest and safest way would be to just concatenate and hash them again (because we know/expect the length of parts to be fixed): `crypto.createHash('sha256').update(uuidBytes).update(hashBytes).digest()`

So, why not just `uuidv4({random: c.createHash('sha256').update(uuidBytes).update(hashBytes).digest().slice(0,16)})`?

Reply:
The native side expects `compilerId` to be a valid uuid (it tries to parse it)

Reply:
Also important is that if DAP isn't applied, `compilerId` stays the same. This is to make sure we don't accidentally introduce any bug in non-DAP use-cases

Reply:
Ah, ok. Now I see your point
updated the code